File: src/main/resources/project/server/modifyApplicationClassLoader.pl
#
#  Copyright 2015 Electric Cloud, Inc.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

=head1 NAME

modifyApplicationClassLoader.pl - a perl library to modify application's ClassLoader behaviour.

=head1 SYNOPSIS

=head1 DESCRIPTION

A perl library that modifies application's ClassLoader behaviour.

=head1 LICENSE

Copyright (c) 2014 Electric Cloud, Inc.
All rights reserved.

=head1 AUTHOR

    ---

=head2 METHODS

=cut

# -------------------------------------------------------------------------
# Includes
# -------------------------------------------------------------------------
use ElectricCommander;
use ElectricCommander::PropMod qw(/myProject/modules);
use WebSphere::WebSphere;
use WebSphere::Util;

use warnings;
use strict;
$| = 1;

# -------------------------------------------------------------------------
# Variables
# -------------------------------------------------------------------------

my $configName        = trim(q($[configurationName]));
my $wsadminAbsPath    = trim(q($[wsadminAbsPath]));
my $applicationName   = trim(q($[applicationName]));
my $loadOrder         = trim(q($[loadOrder]));
my $classLoaderPolicy = trim(q($[classLoaderPolicy]));

# create args array
my @args  = ();
my %props = ();


#get an EC object
my $ec = new ElectricCommander();
$ec->abortOnError(0);

print "Will modify ClassLoader for " . $applicationName . " application. Load Order: $loadOrder, ClassLoader policy: $classLoaderPolicy\n";
my $websphere = new WebSphere::WebSphere( $ec, $configName, $wsadminAbsPath );

my $file = 'modify_application_classloader.py';
my $script = $ec->getProperty("/myProject/wsadmin_scripts/$file")->getNodeText('//value');

$file = $websphere->write_jython_script(
    $file, {},
    augment_filename_with_random_numbers => 1
);

my $shellcmd = $websphere->_create_runfile( $file, @args );
my $escapedCmdLine = $websphere->_mask_password($shellcmd);

print "WSAdmin command line:  $escapedCmdLine\n";
$props{'modifyApplicationClassLoaderLine'} = $escapedCmdLine;
setProperties( $ec, \%props );

#execute command
print `$shellcmd 2>&1`;

#evaluates if exit was successful to mark it as a success or fail the step
if ( $? == SUCCESS ) {
    $ec->setProperty( "/myJobStep/outcome", 'success' );
}
else {
    $ec->setProperty( "/myJobStep/outcome", 'error' );
}

1;

File: src/main/resources/project/wsadmin_scripts/modify_application_classloader.py
#
#  Copyright 2016 Electric Cloud, Inc.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

import sys

applicationName = r'''
$[applicationName]
'''.strip()

loadOrder = r'''
$[loadOrder]
'''.strip()

classLoaderPolicy = r'''
$[classLoaderPolicy]
'''.strip()

dep = AdminConfig.getid('/Deployment:' + applicationName + '/')
if not dep :
    print "No application was found by name provided: %s" % (applicationName)
    sys.exit(1);
print "Application: " + dep

depObject = AdminConfig.showAttribute(dep, 'deployedObject')
if not depObject:
    print "Deployed Object wasn't found"
    sys.exit(1);
print "DeployerObject: " + depObject;

classldr = AdminConfig.showAttribute(depObject, 'classloader')
if not classldr:
    print "ClassLoader %s wasn't found" % (classldr)
    sys.exit(1);
print "ClassLoader: " + classldr

print AdminConfig.showall(classldr)

# load order PARENT_LAST PARENT_FIRST
AdminConfig.modify(classldr, [['mode', loadOrder]])
# loading policy - SINGLE, MULTIPLE
AdminConfig.modify(depObject, [['warClassLoaderPolicy', classLoaderPolicy]])

AdminConfig.save()

# Obtain deployment manager MBean
dm = AdminControl.queryNames('type=DeploymentManager,*')

# Synchronization of configuration changes is only required in network deployment.not in standalone server environment.
if dm:
    print 'Synchronizing configuration repository with nodes. Please wait...'
    nodes=AdminControl.invoke(dm, "syncActiveNodes", "true")
    print 'The following nodes have been synchronized:'+str(nodes)
else:
    print 'Standalone server, no nodes to sync'

